eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ExternalDependencyExposedCheck.java

```diff
@@ -0,0 +1,236 @@
+    }
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.IMPORT,
+            TokenTypes.METHOD_DEF
+        };
+    }
+
+    @Override
+    public void visitToken(DetailAST token) {
+        switch (token.getType()) {
+            case TokenTypes.IMPORT:
+                addImportedClassPath(token);
+                break;
+            case TokenTypes.METHOD_DEF:
+                checkNoExternalDependencyExposed(token);
+                break;
+            default:
+                // Checkstyle complains if there's no default block in switch
+                break;
+        }
+    }
+
+    /**
+     * Add all imported classes into a map, key is the name of class and value is the full package path of class.
+     *
+     * @param token the IMPORT AST node
+     */
+    private void addImportedClassPath(DetailAST token) {
+        final String importClassPath = FullIdent.createFullIdentBelow(token).getText();
+        final String className = importClassPath.substring(importClassPath.lastIndexOf(DOT) + 1);
+        classPathMap.put(className, importClassPath);
+    }
+
+    /**
+     * Checks for external dependency, log the error if it is an invalid external dependency.
+     *
+     * @param methodDefToken METHOD_DEF AST node
+     */
+    private void checkNoExternalDependencyExposed(DetailAST methodDefToken) {
+        final DetailAST modifiersToken = methodDefToken.findFirstToken(TokenTypes.MODIFIERS);
+        if (modifiersToken == null) {
+            return;
+        }
+
+        AccessModifier accessModifier = CheckUtil.getAccessModifierFromModifiersToken(modifiersToken);
+        if (!accessModifier.equals(AccessModifier.PUBLIC) && !accessModifier.equals(AccessModifier.PROTECTED)) {
+            return;
+        }
+
+        DetailAST typeToken = methodDefToken.findFirstToken(TokenTypes.TYPE);
+        if (typeToken != null) {
+            getInvalidReturnTypes(typeToken).forEach(
+                (token, returnTypeName) -> log(token, String.format(EXTERNAL_DEPENDENCY_ERROR, returnTypeName)));
+        }
+
+        DetailAST parametersToken = methodDefToken.findFirstToken(TokenTypes.PARAMETERS);
+        if (parametersToken != null) {
+            getInvalidParameterTypes(parametersToken).forEach(
+                (token, returnTypeName) -> log(token, String.format(EXTERNAL_DEPENDENCY_ERROR, returnTypeName)));;
+        }
+    }
+
+    /**
+     * Get invalid return types from a given TYPE node.
+     *
+     * @param typeToken TYPE AST node
+     * @return a map that maps the invalid TYPE node and the type name.
+     */
+    private Map<DetailAST, String> getInvalidReturnTypes(DetailAST typeToken) {
+        Map<DetailAST, String> invalidReturnTypeMap = new HashMap<>();
```

Comment:
A return type always has an "IDENT" but what it might not have are TYPE_ARGUMENTS. You're duplicating the logic for checking the IDENT.

Example:  `Mono<String> foo() { }`
1. Has a token type IDENT with value "Mono"
1. Also has a TYPE_ARGUMENTS child node with a TYPE_PARAMETER.

Example: `EventHubProducer createProducer()`
1. Has a token type IDENT with value "EventHubProducer".